### PR TITLE
Delete favorite routes and stops

### LINKED ISSUE #379: Swipe To Delete Favorites On My Buses
In PVTrAck 1.x, one could swipe left on their favorites on MyBuses and be greeted with a way to delete them, like so:
merp

PVTrAck 2.x doesn't have this yet, and would be a nice little addition.  Doing so will require small changes to the MyBuses view and component, and will familiarize you with some of Ionic's HTML directives, event handing in TypeScript, and our custom Favorites schemas.

### PR DESCRIPTION
**Pull request by mboneil10**
Closes #379

### CODE REVIEW

**Comment by mboneil10** [comment]
Doesn't remove favorite routes and stops from the view as it does with trips. So, I can swipe to delete both stops and routes, but the view doesn't update unless I refresh. I used the instance-variable array for both routes and stops. What gives (@akaplo)? I'll look into this later.

**Comment by mboneil10** [comment]
Oops I'll update the tests so there isn't a CI fail! AGGGHHH

**Comment by mboneil10** [comment]
And yes, needs tests still. I put this PR up just so other people could help me figure out why the instance-variable arrays don't update the view automatically.

File: src/pages/favorites/favorites.component.ts [hunk 1]
```diff
@@ -164,6 +167,14 @@ export class FavoritesComponent {
   }
   deleteTrip(trip): void {
     _.remove(this.trips, {name: trip.name});
-    this.tripSvc.deleteTrip(trip);
+    this.faveTripSvc.deleteTrip(trip);
+  }
+  removeRoute(route): void {
+    _.remove(this.routes, {name: route.name});
```
Comment (akaplo) [inline src/pages/favorites/favorites.component.ts]: So, you're asking [Lodash](https://lodash.com/docs/4.17.4#remove) to `remove` entries in the `this.routes` array whose `name` property matches the `name` property of the route passed to this function.  

The problem here is that routes don't have a `name` property!  This line should be
`_.remove(this.routes, {RouteId: route.RouteId})`. 

The reason why the [FavoriteRouteService function](https://github.com/umts/pvta-multiplatform/blob/master/src/providers/favorite-route.service.ts#L56) call works is that it's properly removing based on `RouteId`.

File: src/pages/favorites/favorites.component.ts [hunk 2]
```diff
@@ -164,6 +167,14 @@ export class FavoritesComponent {
...
+    this.faveRouteSvc.remove(route);
+  }
+  removeStop(stopId): void {
+    _.remove(this.stops, {stopId: stopId});
```
Comment (akaplo) [inline src/pages/favorites/favorites.component.ts]: Similar issue here as above, but this one's more subtle - `stopId` is the name of the param that's passed to this function, but a stop object doesn't have a `stopId` property - it's called ` StopId` (note the capital `S`).  Woof.

**Review by akaplo — CHANGES_REQUESTED** [review]
2 quick changes.

File: src/pages/favorites/favorites.component.ts (hunk 1 above)
Comment (mboneil10) [reply src/pages/favorites/favorites.component.ts]: Ohhh. I saw `routeToRemove: Route`, and thought that `Route` would just be the name of the route. I was wrong. Great!

File: src/pages/favorites/favorites.component.ts (hunk 2 above)
Comment (mboneil10) [reply src/pages/favorites/favorites.component.ts]: WAT

Comment (mboneil10) [reply src/pages/favorites/favorites.component.ts]: WHY

Comment (mboneil10) [reply src/pages/favorites/favorites.component.ts]: (I'll be looking into this, no need to actually answer the above question of why)

File: src/pages/favorites/favorites.component.ts (hunk 1 above)
Comment (mboneil10) [reply src/pages/favorites/favorites.component.ts]: Wow that makes so much more sense to me. Thank you.

**Comment by mboneil10** [comment]
Next, tests

File: src/pages/favorites/favorites.component.ts (hunk 1 above)
Comment (akaplo) [reply src/pages/favorites/favorites.component.ts]: Hooray! You're welcome.

Comment (akaplo) [reply src/pages/favorites/favorites.component.ts]: FYI the structure of all the Avail objects (like Route, RouteDetails, Stop, StopDeparture, etc) is currently best found on [the Swagger doc](http://bustracker.pvta.com/InfoPoint/swagger/ui/index).

**Comment by akaplo** [comment]
Yup! If you have testing questions, totes throw them my way.

File: src/pages/favorites/favorites.component.spec.ts [hunk 3]
```diff
@@ -71,4 +71,37 @@ describe('Favorites Component', () => {
...
+        RouteId: 20035,
+        GoogleDescription: 'Campus Shuttle Southbound',
+        ShortName: '35',
+        RouteAbbreviation: '35',
+        Color: '00467E'
+      }
+    });
+    it('removes an item from the routes array', () => {
+      // Set the class-wide this.routes array in the component
+      component.routes = [ this.route ];
```
Comment (mboneil10) [inline src/pages/favorites/favorites.component.spec.ts:87]: Ohhhhh! That's how you do it. I was trying to call `this.routes`.

File: src/pages/favorites/favorites.component.spec.ts [hunk 4]
```diff
@@ -71,4 +71,37 @@ describe('Favorites Component', () => {
...
   });
+  describe('removeRoute', () => {
+    // Make an arbitrary route object available to each test
+    beforeAll(() => {
+      this.route = {
+        RouteId: 20035,
+        GoogleDescription: 'Campus Shuttle Southbound',
+        ShortName: '35',
+        RouteAbbreviation: '35',
+        Color: '00467E'
```
Comment (mboneil10) [inline src/pages/favorites/favorites.component.spec.ts]: Are all of these necessary? If so, is there any way to make a factory in ionic/angular?

File: src/pages/favorites/favorites.component.spec.ts [hunk 5]
```diff
@@ -71,4 +71,37 @@ describe('Favorites Component', () => {
...
+      // https://stackoverflow.com/questions/35733846/how-to-spy-a-service-call-in-angular2
+      let mockFavRouteSvc = fixture.debugElement.injector.get(FavoriteRouteService);
+      // Set the class-wide variable again (component isn't available in beforeAll?)
+      component.routes = [ this.route ];
+      // Insert ourselves into the FavoriteRouteService and
+      // listen specifically for stuff related to remove()
+      spyOn(mockFavRouteSvc, 'remove');
+      // Call our function
+      component.removeRoute(this.route);
+      expect(mockFavRouteSvc.remove).toHaveBeenCalledWith(this.route)
```
Comment (mboneil10) [inline src/pages/favorites/favorites.component.spec.ts]: Do you need semicolons?

File: src/pages/favorites/favorites.component.spec.ts [hunk 6]
```diff
@@ -71,4 +71,37 @@ describe('Favorites Component', () => {
...
+      expect(component.routes.length).toEqual(1);
+      // Call the function we're testing`
+      component.removeRoute(this.route);
+      expect(component.routes.length).toEqual(0);
+    });
+    it('calls the FavoriteRouteService\'s remove() function', () => {
+      // Obtain a reference to a the service
+      // https://stackoverflow.com/questions/35733846/how-to-spy-a-service-call-in-angular2
+      let mockFavRouteSvc = fixture.debugElement.injector.get(FavoriteRouteService);
+      // Set the class-wide variable again (component isn't available in beforeAll?)
```
Comment (mboneil10) [inline src/pages/favorites/favorites.component.spec.ts]: I wonder why. Hm

File: src/pages/favorites/favorites.component.spec.ts (hunk 4 above)
Comment (mboneil10) [reply src/pages/favorites/favorites.component.spec.ts]: And where did you get them? Swagger?

Comment (mboneil10) [reply src/pages/favorites/favorites.component.spec.ts]: Oh, I found it in [favorite route service](https://github.com/umts/pvta-multiplatform/blob/master/src/providers/favorite-route.service.ts). Isn't this technically a favorite route then? Or do all routes have these properties?

Comment (akaplo) [reply src/pages/favorites/favorites.component.spec.ts]: This is indeed a favorite route - in this case specifically, I threw a debugger into the FavoritesComponent and copied one of _my_ favorite routes when it was loaded onto the page.

I don't know anything about factories here, but I hope so!  For this test, we technically could use any object with the `RouteId` property set, as that's all the code we're testing is looking for.

File: src/pages/favorites/favorites.component.spec.ts (hunk 3 above)
Comment (akaplo) [reply src/pages/favorites/favorites.component.spec.ts:87]: 👍 scope is difficult to understand in the tests - in general:
- Assign `this.xyz` variables in `beforeEach` or `beforeAll` and access them in `it()` blocks.  These variables have nothing to do with the `this.zyx` variables in the component.
- Set and get `this.zyx` variables that we're used to seeing in the component via `component.zyx` in the `it()` blocks.

File: src/pages/favorites/favorites.component.spec.ts (hunk 5 above)
Comment (akaplo) [reply src/pages/favorites/favorites.component.spec.ts]: Yes indeed.

File: src/pages/favorites/favorites.component.spec.ts (hunk 6 above)
Comment (akaplo) [reply src/pages/favorites/favorites.component.spec.ts]: I'd assume it has to do funky scoping or timing - like maybe the `beforeEach()` that sets it doesn't always run before this one, or something of that sort.

File: src/pages/favorites/favorites.component.spec.ts (hunk 5 above)
Comment (mboneil10) [reply src/pages/favorites/favorites.component.spec.ts]: I will add them in then

File: src/pages/favorites/favorites.component.spec.ts (hunk 6 above)
Comment (mboneil10) [reply src/pages/favorites/favorites.component.spec.ts]: Right. I was thinking it was a timing thing, too.

File: src/pages/favorites/favorites.component.spec.ts (hunk 4 above)
Comment (mboneil10) [reply src/pages/favorites/favorites.component.spec.ts]: I'll look into factories... after this. Seems like a different issue all together. I can add it.

**Comment by mboneil10** [comment]
I'll follow the basic structure of what @akaplo laid out for the tests. It will only be two or three tests, so no big deal if we have to rewrite them later... I hope.

File: src/pages/favorites/favorites.component.spec.ts [hunk 7]
```diff
@@ -67,4 +71,80 @@ describe('Favorites Component', () => {
...
+      // Insert ourselves into the FavoriteRouteService and
+      // listen specifically for stuff related to remove()
+      spyOn(mockFaveRouteSvc, 'remove');
+      // Call our function
+      component.removeRoute(this.route);
+      expect(mockFaveRouteSvc.remove).toHaveBeenCalledWith(this.route);
+    });
+  });
+  describe('removeStop', () => {
+    // Make an arbitrary route object available to each test
```
Comment (akaplo) [inline src/pages/favorites/favorites.component.spec.ts]: arbitrary stop object

File: src/pages/favorites/favorites.component.spec.ts [hunk 8]
```diff
@@ -67,4 +71,80 @@ describe('Favorites Component', () => {
...
+    it("calls the FavoriteStopService's remove() function", () => {
+      let mockFaveStopSvc = fixture.debugElement.injector.get(FavoriteStopService);
+      component.stops = [ this.stop ];
+      spyOn(mockFaveStopSvc, 'remove');
+      component.removeStop(this.stop);
+      expect(mockFaveStopSvc.remove).toHaveBeenCalledWith(this.stop);
+    });
+  });
+  describe('deleteTrip', () => {
+    // Make an arbitrary route object available to each test
```
Comment (akaplo) [inline src/pages/favorites/favorites.component.spec.ts]: arbitrary trip 🥇

**Review by akaplo — APPROVED** [review]

**Comment by akaplo** [comment]
Hooray! Great job with this.

File: src/pages/favorites/favorites.component.spec.ts (hunk 8 above)
Comment (mboneil10) [reply src/pages/favorites/favorites.component.spec.ts]: OOp

File: src/pages/favorites/favorites.component.spec.ts (hunk 7 above)
Comment (mboneil10) [reply src/pages/favorites/favorites.component.spec.ts]: That's what copy and paste does to you, man.

Comment (mboneil10) [reply src/pages/favorites/favorites.component.spec.ts]: Never again.

File: src/pages/favorites/favorites.component.spec.ts (hunk 8 above)
Comment (mboneil10) [reply src/pages/favorites/favorites.component.spec.ts]: I just deleted those comments and left the routes comments.

**Pull request merged**